File: leetcode/LC1/array/binary_search/0074_search_in_sorted_2D_matrix.cpp
#include <climits>
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <unordered_map>
using namespace std;


class Solution {
    // O(logmn)
    bool searchMatrix(vector<vector<int>> matrix, int target) {
        int r = matrix.size();
        int c = matrix[0].size();
        int s = 0;
        int l = r*c - 1;
        
        int mid = -1;
        // note less than or equal to
        while(s <= l){
            mid =  (l -s)/2 + s;
            // mid / c will give number of rows passed and mid %c will give number of columns passed
            if(matrix[mid/c][mid%c] == target){
                return true;
            }else if(matrix[mid/c][mid%c] > target){
                l = mid  - 1;
            }else{
                s = mid + 1;
            }  
        }
        return false;
    }
};

File: leetcode/LC/array/binary_search/0074_search_insorted_2D_matrix.py
from typing import List

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        row = len(matrix)
        col = len(matrix[0])
        
        start = 0
        end = row*col - 1
        
        while(start <= end):
            mid = (start + end)//2
            r = mid//col
            c = mid%col
            if(matrix[r][c] == target):
                return True
            if(matrix[r][c] < target):
                start = mid +1
            else:
                end = mid - 1
        return False
        